TMessagesProj/src/main/java/org/telegram/messenger/MediaController.java

```diff
@@ -3913,99 +3923,161 @@ public static void saveFile(String fullPath, Context context, final int type, fi
+                                if (mimeType.startsWith("video")) {
+                                    selectedType = 1;
+                                }
+                            }
+                            if (selectedType == 0) {
+                                uriToInsert = MediaStore.Images.Media.getContentUri(MediaStore.VOLUME_EXTERNAL_PRIMARY);
+                                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+                                    File dirDest = new File(Environment.DIRECTORY_PICTURES, "Telegram");
+                                    contentValues.put(MediaStore.MediaColumns.RELATIVE_PATH, dirDest + File.separator);
+                                }
+                                contentValues.put(MediaStore.Images.Media.DISPLAY_NAME, AndroidUtilities.generateFileName(0, extension));
+                                contentValues.put(MediaStore.Images.Media.MIME_TYPE, mimeType);
+                            } else if (selectedType == 1) {
+                                File dirDest = new File(Environment.DIRECTORY_MOVIES, "Telegram");
+                                contentValues.put(MediaStore.MediaColumns.RELATIVE_PATH, dirDest + File.separator);
+                                uriToInsert = MediaStore.Video.Media.getContentUri(MediaStore.VOLUME_EXTERNAL_PRIMARY);
+                                contentValues.put(MediaStore.Video.Media.DISPLAY_NAME, AndroidUtilities.generateFileName(1, extension));
+                            } else if (selectedType == 2) {
+                                File dirDest = new File(Environment.DIRECTORY_DOWNLOADS, "Telegram");
+                                contentValues.put(MediaStore.MediaColumns.RELATIVE_PATH, dirDest + File.separator);
+                                uriToInsert = MediaStore.Downloads.getContentUri(MediaStore.VOLUME_EXTERNAL_PRIMARY);
+                                contentValues.put(MediaStore.Downloads.DISPLAY_NAME, sourceFile.getName());
+                            } else {
+                                File dirDest = new File(Environment.DIRECTORY_MUSIC, "Telegram");
+                                contentValues.put(MediaStore.MediaColumns.RELATIVE_PATH, dirDest + File.separator);
+                                uriToInsert = MediaStore.Audio.Media.getContentUri(MediaStore.VOLUME_EXTERNAL_PRIMARY);
+                                contentValues.put(MediaStore.Audio.Media.DISPLAY_NAME, sourceFile.getName());
+                            }
+
+                            contentValues.put(MediaStore.MediaColumns.MIME_TYPE, mimeType);
+
+                            Uri dstUri = context.getContentResolver().insert(uriToInsert, contentValues);
+                            if (dstUri != null) {
+                                FileInputStream fileInputStream = new FileInputStream(sourceFile);
+                                OutputStream outputStream = context.getContentResolver().openOutputStream(dstUri);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *outputStream*. Currently, there are execution paths that do not contain closure statements, for example, when AndroidUtilities.copyFile throws an exception. Ensure *outputStream* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).